core/modules/qdisp/testQDisp.cc

```diff
@@ -0,0 +1,412 @@
+	std::vector<char> _vect;
+	std::string _msg;
+	int _code;
+	bool _finished;
+};
+
+// Start adds 'copies' number of test requests that each sleep for 'millisecs' time
+// before signaling to 'ex' that they are done.
+// Returns time to complete in seconds.
+double executiveTest(qdisp::Executive &ex, SequentialInt &sequence, SequentialInt &chunkId, std::string &millisecs, int copies) {
+	// Test class Executive::add
+	// Modeled after ccontrol::UserQuery::submit()
+	ResourceUnit ru;
+	std::string chunkResultName = "mock";
+	boost::shared_ptr<rproc::InfileMerger> infileMerger;
+	boost::shared_ptr<ChunkMsgReceiverMock> cmr = ChunkMsgReceiverMock::newInstance(chunkId.get());
+	ccontrol::MergingRequester::Ptr mr = boost::make_shared<ccontrol::MergingRequester>(cmr, infileMerger, chunkResultName);
+	std::string msg = millisecs;
+	boost::chrono::system_clock::time_point start = boost::chrono::system_clock::now();
+	std::vector<qdisp::Executive::Spec> s(copies);
+	for (int j=0; j < copies; ++j) {
+		s[j].resource = ru;
+		s[j].request = msg;
+		s[j].requester = mr;
+		ex.add(sequence.incr(), s[j]);
+	}
+	ex.join();
+	boost::chrono::duration<double> secs = boost::chrono::system_clock::now() - start;
+	double ret = secs.count();
+	return ret;
+}
+
+class WorkerTest {
+public:
+	double _retVal;
+	void operator()(qdisp::Executive &ex, SequentialInt &sequence, SequentialInt &chunkId, std::string &millisecs, int copies) {
+		_retVal = executiveTest(ex, sequence, chunkId, millisecs, copies);
+	}
+};
+
+// This function is run in a separate thread to fail the test if it takes too long
+// for the jobs to complete.
+void timeoutFunc(qdisp::Executive &ex, int jobs, int millisecs) {
+	LOGF_INFO("timeoutFunc");
+	boost::this_thread::sleep(boost::posix_time::milliseconds(millisecs));
+	LOGF_INFO("timeoutFunc sleep over %1% empty=%2%" % millisecs % ex._empty.get());
+	BOOST_REQUIRE(ex._empty.get() == true);
+}
+
+BOOST_AUTO_TEST_CASE(Executive) {
+	LOGF_INFO("Executive test 1");
+	// Modeled after ccontrol::UserQuery::submit()
+	std::string str = qdisp::Executive::Config::getMockStr();
+	qdisp::Executive::Config::Ptr conf = boost::make_shared<qdisp::Executive::Config>(str);
+	boost::shared_ptr<qdisp::MessageStore> ms = boost::make_shared<qdisp::MessageStore>();
+	qdisp::Executive ex(conf, ms);
+	SequentialInt sequence(0);
+	SequentialInt chunkId(1234);
+	int jobs = 0;
+	// test single instance
+	int millisInt = 400;
+	std::string millis(boost::lexical_cast<std::string>(millisInt));
+	double duration = executiveTest(ex, sequence, chunkId, millis, 1);
+	++jobs;
+	LOGF_INFO("duration single %1%" % duration);
+	BOOST_CHECK(ex._empty.get() == true);
+	// test adding 4 jobs
+	duration = executiveTest(ex, sequence, chunkId, millis, 4);
+	jobs += 4;
+	LOGF_INFO("duration four %1%" % duration);
+	BOOST_CHECK(ex._empty.get() == true);
+	// test multiple threads
+	int threadCount = 10;
+	int jobsPerThread = 2;
+	std::vector<WorkerTest> workers(threadCount);
```

Comment:
Wait, so you are starting a bunch of 'worker' threads that each add 1 or 4 jobs and then join on a single Executive? Executive::add() is not thread-safe or re-entrant. Executive::join() wasn't supposed to be either (though it might be). The intent of an Executive is to have a single manager thread create an instance, add a bunch of things to it, and then wait for things to finish. There should only be one manager thread. Other threads may peek into an Executive object's state. I feel like there is a much more complex usage being tested here. 
We guarantee that Executive never has multiple threads  calling add() or join(), because it is only responsible for dispatch for a single user query, and query analysis/generation is single-threaded. There shouldn't be any benefit(?) to performing add() with parallel threads because the main task of add() is to kick-off xrootd calls (which are async, non-blocking, and using an internal thread pool). The xrootd calls are not supposed to block for any reason.